File: LongestPalindrome.py
#All TC passed on leetcode


class Solution:
    def longestPalindrome(self, s: str) -> str:

        #Here we use 2 pointer approach. We iterate through the string for each char we check char 
        #on either side are equal or not.
        #time complexity - O(n^2)
        #space complexity - O(1)
        start = 0
        end = 0
        for i in range(len(s)):
            low = i-1
            high = i+1
            l, h = self.check(s, low, high)
            if h-l>end-start:
                start = l
                end = h

            low = i
            high = i+1
            l, h = self.check(s, low, high)
            if h-l>end-start:
                start = l
                end = h
        
        return s[start:end+1]


    def check(self, s, l, h):
        n = len(s)
        while l>=0 and h<n and s[l]==s[h]:
            l-=1
            h+=1
        l+=1
        h-=1
        return l, h


#-------------------------------------------------OR--------------------------------------------------
        
        
        #Here we use dp approach. We compare string chars from front and back 
        #time complexity - O(n^2)
        #space complexity - O(n^2)
        n = len(s)
        dp = [[False]*n for i in range(n)]
        start = 0
        end = 0

        for i in range(n):
            for j in range(i+1):
                if (s[i]==s[j]) and (i-j<2 or dp[i-1][j+1]):
                    dp[i][j] = True
                    if i-j>end-start:
                        start = j
                        end = i 
                else:
                    dp[i][j] = False
        return s[start:end+1]

        

File: longestPalindrome.cpp
// Time complexity: O(n^2)
// Space complexity: O(n^2)
/*
we use the row and cols as the end and start point of the substring 
row -> i -> 0 to n
col -> j -> 0 to i

i == j => true and we update the max and indices 
if s[i] == s[j]
1. Its got letters inbetween and we can find the solution by looking at dp[i-1][j+1]
2. Or its true if the letter intween are 0 or 1 => i.e aba or aa scenarios 

*/
class Solution {
public:
    string longestPalindrome(string s) {
        if(s.length() < 2)
            return s;
        int n = s.length();
        int max = -1, start, end;
        vector<vector<bool>> dp (n, vector<bool> (n, false));
        for(int i = 0; i<n; i++) {
            for(int j = 0; j <= i; j++) {
                // i - j - 1 is for aba or aa scenario
                if (s[i] == s[j] && (i-j-1 < 1 || dp[i-1][j+1])) {
                    dp[i][j] = true;
                    if(i-j+1 > max) {
                        max = i-j+1;
                        start = j;
                        end = i;
                    }
                }
            }
        }
        return s.substr(start, max);
    }
};


// Approach 2: two pointer 
// Time complexity: O(n^2)
// Space complexity: O(1)

/*
iterate over every letter and explode the boundaries and get the maximum palindrome
Re-do the same thing but now take pairs of letter which are the same
*/

class Solution {
public:
    int resultStart, max = -1;
    string longestPalindrome(string s) {
        if(s.length() < 2)
            return s;

        for(int i =0; i < s.length(); i++) {
            expandAround (s, i, i);
            if(i+1 < s.length() && s[i] == s[i+1])
                expandAround (s, i, i+1);
        }

        return s.substr(resultStart, max);
    }
private:
    void expandAround(string s, int start, int end) {
        while (start >= 0 && end < s.length() && s[start] == s[end]) {
            start --; 
            end ++;
        }
        start ++;
        end --; 
        if(end - start + 1 > max) {
            max = end - start + 1;
            resultStart = start;
        }
    }
};